pkg/networkserver/redis/application_uplink_queue.go

```diff
@@ -207,20 +205,15 @@ func (q *ApplicationUplinkQueue) processMessages(
 			return err
 		}
 	}
-	pipeliner := q.redis.Pipeline()
+	processedIDs := make([]string, 0, len(msgs))
 	for _, batch := range batches {
 		if err := f(batch.ctx, batch.uplinks); err != nil {
 			log.FromContext(ctx).WithError(err).Warn("Failed to process uplink batch")
 			continue // Do not confirm messages that failed to process.
 		}
-
-		pipeliner.XAck(ctx, q.streamID, q.groupID, batch.confirmIDs...)
-		pipeliner.XDel(ctx, q.streamID, batch.confirmIDs...)
-	}
-	if _, err := pipeliner.Exec(ctx); err != nil {
-		return ttnredis.ConvertError(err)
+		processedIDs = append(processedIDs, batch.confirmIDs...)
 	}
-	return nil
+	return ack(processedIDs...)
```

Comment:
nit: Maybe we should convert the Redis error here

Reply:
You are correct, I will add it into the `RangeStreams` code.